observatory-dags/observatory/dags/telescopes/oapen_metadata.py

```diff
@@ -0,0 +1,227 @@
+
+
+# TODO
+# isbn have '-' sometimes, so not an integer, leave in?
+
+class OapenMetadataRelease(StreamRelease):
+    def __init__(self, dag_id: str, start_date: pendulum.Pendulum, end_date: pendulum.Pendulum, first_release: bool):
+        super().__init__(dag_id, start_date, end_date, first_release)
+
+    @property
+    def csv_path(self) -> str:
+        """ Path to store the original oapen metadata csv file"""
+        return os.path.join(self.download_folder, 'oapen_metadata.csv')
+
+    @property
+    def transform_path(self) -> str:
+        """ Path to store the transformed oapen metadata file"""
+        return os.path.join(self.transform_folder, 'metadata.jsonl')
+
+    def download(self) -> bool:
+        """ Download Oapen metadata CSV.
+        :return: True if download is successful
+        """
+        logging.info('Downloading csv')
+        headers = {
+            'User-Agent': f'{get_ao_user_agent()}'
+        }
+        response = retry_session().get(OapenMetadataTelescope.CSV_URL, headers=headers)
+        if response.status_code == 200:
+            with open(self.csv_path, 'w') as f:
+                f.write(response.content.decode('utf-8'))
+            logging.info('Download csv successful')
+            return True
+        else:
+            raise AirflowException(f'Download csv unsuccessful, {response.text}')
+
+    def transform(self):
+        """ Transform the oapen metadata csv file by storing in a jsonl format and restructuring lists/dicts
+        :return: None
+        """
+        with open(self.csv_path, 'r') as f:
+            csv_entries = [{k: v for k, v in row.items()} for row in csv.DictReader(f, skipinitialspace=True)]
+
+        # restructure entries dict
+        nested_fields = set([key.rsplit('.', 1)[0] for key in csv_entries[0].keys() if len(key.rsplit('.', 1)) > 1])
+        list_fields = {'oapen.grant.number', 'oapen.grant.acronym', 'oapen.relation.hasChapter_dc.title',
+                       'dc.subject.classification', 'oapen.grant.program', 'oapen.redirect', 'oapen.imprint',
+                       'dc.subject.other', 'oapen.notes', 'dc.title', 'collection', 'dc.relation.ispartofseries',
+                       'BITSTREAM Download URL', 'BITSTREAM ISBN', 'oapen.collection', 'dc.contributor.author',
+                       'oapen.remark.public', 'oapen.relation.isPublisherOf', 'dc.contributor.other',
+                       'dc.contributor.editor', 'dc.relation.isreplacedbydouble', 'dc.date.issued',
+                       'dc.description.abstract', 'BITSTREAM Webshop URL', 'dc.type', 'dc.identifier.isbn',
+                       'dc.description.provenance', 'oapen.grant.project', 'oapen.relation.isbn', 'dc.identifier',
+                       'dc.date.accessioned', 'BITSTREAM License', 'oapen.relation.isFundedBy_grantor.name',
+                       'dc.relation.isnodouble', 'dc.language', 'grantor.number'}
+        entries = change_keys(csv_entries, convert, nested_fields, list_fields)
```

Comment:
It would be good to add a description to the docstrings that describes how the dataset is transformed, including an example of what a row of the dataset looks like originally and what it looks like after transformation.

Reply:
I agree it would be helfpul to add more info on what's going on during transform. In this case the rows are quite lengthy. Maybe I can add an example of a row before vs after in the readthedocs part of this telescope instead, what do you think? These are the before and after examples for 1 row:

Before
```
{'id': 'af17b762-8a0c-4719-aa9c-7f5083a0ba52',
'collection': '20.500.12657/6',
'BITSTREAM Download URL': 'https://library.oapen.org/bitstream/20.500.12657/29427/1/academia-LLS.2018.10_160x240.pdf',
'BITSTREAM ISBN': '9789401452113',
'BITSTREAM License': 'http://creativecommons.org/licenses/by-nd/3.0/',
'BITSTREAM Webshop URL': '',
'dc.contributor.advisor': '',
'dc.contributor.author': 'Missinne  , Lut||Grave , Jaap',
'dc.contributor.editor': '',
'dc.contributor.other': '',
'dc.date.accessioned': '2018-09-12 23:55||2019-04-18 16:22:31||2020-04-01T12:27:51Z',
'dc.date.available': '2020-04-01T12:27:51Z',
'dc.date.issued': '2018',
'dc.description': '',
'dc.description.abstract': 'The abstract',
'dc.description.provenance': "Created by dorien on 2018-09-12 14:55:07||Edited by lotte on 2019-04-18 16:22:31||Made available in DSpace on 2020-04-01T12:27:51Z (GMT). No. of bitstreams: 1\nacademia-LLS.2018.10_160x240.pdf: 13825548 bytes,
checksum: a60085ebcce61de9bc7d22efc2924dce (MD5)\n  Previous issue date: 2018||Item created via OAI harvest from source: https://dataprovider.oapen.org/cgi/oapen on 2020-04-01T12:27:51Z (GMT).  Item's OAI Record identifier: oai:oapen:1000509",
'dc.description.version': '',
'dc.identifier': '1000509',
'dc.identifier.isbn': '',
'dc.identifier.issn': '',
'dc.identifier.uri': 'http://library.oapen.org/handle/20.500.12657/29427',
'dc.identifier.urlwebshop': '',
'dc.language': 'Dutch; Flemish[dut]||German[ger]',
'dc.relation.isnodouble': '',
'dc.relation.ispartofseries': 'Lage Landen Studies',
'dc.relation.isreplacedbydouble': '',
'dc.rights.uri': '',
'dc.source': '',
'dc.subject.classification': 'bic Book Industry Communication::C Language::CF linguistics::CFP Translation & interpretation||bic Book Industry Communication::D Literature & literary studies||bic Book Industry Communication::D Literature & literary studies::DS Literature: history & criticism::DSK Literary studies: fiction, novelists & prose writers||bic Book Industry Communication::F Fiction & related items',
'dc.subject.other': 'Dutch fiction||fiction in translation||translation||literary transfer||cultural transfer',
'dc.title': 'Tussen twee stoelen, tussen twee vuren',
'dc.title.alternative': 'Nederlandse literatuur op weg naar de buitenlandse lezer',
'dc.type': 'book',
'dcterms.abstract': '',
'eperson.firstname': '',
'eperson.language': '',
'eperson.lastname': '',
'eperson.phone': '',
'grantor.acronym': '',
'grantor.doi': '',
'grantor.jurisdiction': '',
'grantor.name': '',
'grantor.number': '',
'oapen.abstract.otherlanguage': 'Abstract other language',
'oapen.autodoi': '',
'oapen.chapternumber': '',
'oapen.collection': '',
'oapen.description.otherlanguage': '',
'oapen.embargo': '',
'oapen.grant.acronym': '',
'oapen.grant.number': '',
'oapen.grant.program': '',
'oapen.grant.project': '',
'oapen.identifier': '',
'oapen.identifier.doi': '',
'oapen.identifier.isbn': '',
'oapen.imprint': '',
'oapen.notes': '',
'oapen.pages': '222',
'oapen.place.publication': 'Gent',
'oapen.redirect': '',
'oapen.relation.funds': '',
'oapen.relation.hasChapter': '',
'oapen.relation.hasChapter_dc.title': '',
'oapen.relation.isFundedBy': '',
'oapen.relation.isFundedBy_grantor.name': 'Internationale Vereniging voor Neerlandistiek',
'oapen.relation.isPartOfBook': '',
'oapen.relation.isPartOfBook_dc.title': '',
'oapen.relation.isPublishedBy': '',
'oapen.relation.isPublishedBy_publisher.name': 'Academia Press',
'oapen.relation.isPublisherOf': '',
'oapen.relation.isbn': '',
'oapen.remark.private': '',
'oapen.remark.public': '21-7-2020 - No DOI registered in CrossRef for ISBN 9789401452113',
'oapen.series.number': '10',
'publisher.country': '',
'publisher.name': '',
'publisher.peerreviewpolicy': '',
'publisher.website': ''}
```

After
```
 {'id': 'af17b762-8a0c-4719-aa9c-7f5083a0ba52', 
'collection': ['20.500.12657/6'], 
'BITSTREAM_Download_URL': ['https://library.oapen.org/bitstream/20.500.12657/29427/1/academia-LLS.2018.10_160x240.pdf'], 
'BITSTREAM_ISBN': ['9789401452113'], 
'BITSTREAM_License': ['http://creativecommons.org/licenses/by-nd/3.0/'], 
'dc': {'contributor': {'author': ['Missinne  , Lut', 'Grave , Jaap']}, 
       'date': {'accessioned': ['2018-09-12 23:55', '2019-04-18 16:22:31', '2020-04-01T12:27:51Z'], 
                'available': '2020-04-01T12:27:51Z', 'issued': ['2018-01-01']}, 
       'description': {'abstract': ['The abstract.'], 
                       'provenance': ['Created by dorien on 2018-09-12 14:55:07', 
                                      'Edited by lotte on 2019-04-18 16:22:31', 
                                      'Made available in DSpace on 2020-04-01T12:27:51Z (GMT). No. of bitstreams: 1\nacademia-LLS.2018.10_160x240.pdf: 13825548 bytes, checksum: a60085ebcce61de9bc7d22efc2924dce (MD5)\n  Previous issue date: 2018', 
                                      "Item created via OAI harvest from source: https://dataprovider.oapen.org/cgi/oapen on 2020-04-01T12:27:51Z (GMT).  Item's OAI Record identifier: oai:oapen:1000509"]}, 
       'identifier': {'value': ['1000509'], 
                      'uri': 'http://library.oapen.org/handle/20.500.12657/29427'}, 
       'language': ['Dutch; Flemish[dut]', 'German[ger]'], 
       'relation': {'ispartofseries': ['Lage Landen Studies']}, 
       'subject': {'classification': ['bic Book Industry Communication::C Language::CF linguistics::CFP Translation & interpretation', 
                                      'bic Book Industry Communication::D Literature & literary studies', 
                                      'bic Book Industry Communication::D Literature & literary studies::DS Literature: history & criticism::DSK Literary studies: fiction, novelists & prose writers', 
                                      'bic Book Industry Communication::F Fiction & related items'], 
                   'classification_code': ['CFP', 'D', 'DSK', 'F'], 
                   'other': ['Dutch fiction', 'fiction in translation', 'translation', 'literary transfer', 'cultural transfer']}, 
       'title': {'value': ['Tussen twee stoelen, tussen twee vuren'], 
                 'alternative': 'Nederlandse literatuur op weg naar de buitenlandse lezer'}, 
       'type': ['book']}, 
'oapen': {'abstract': {'otherlanguage': 'Abstract other language.'}, 
          'pages': '222', 'place': {'publication': 'Gent'}, 
          'relation': {'isFundedBy_grantor': {'name': ['Internationale Vereniging voor Neerlandistiek']}, 
                       'isPublishedBy_publisher': {'name': 'Academia Press'}}, 
          'remark': {'public': ['21-7-2020 - No DOI registered in CrossRef for ISBN 9789401452113']}, 
          'series': {'number': '10'}}}
```